test/integration/options/watch.spec.js

```diff
@@ -19,38 +21,156 @@ describe('--watch', function() {
+          touchFile(this.testFile);
+          return sleep(1000);
+        })
+        .then(() => {
+          mocha.kill('SIGINT');
+          return outputPromise;
+        })
+        .then(results => {
+          expect(results, 'to have length', 2);
+        });
+    });
+
+    it('reruns test when file matching extension is touched', function() {
+      const watchedFile = path.join(this.tempDir, 'file.xyz');
+      touchFile(watchedFile);
+      const [mocha, outputPromise] = runMochaJSONWatchAsync(
+        this.testFile,
+        ['--extension', 'xyz,js'],
+        {
+          cwd: this.tempDir
+        }
       );
 
-      setTimeout(function() {
-        // Kill the child process
-        mocha.kill('SIGINT');
-      }, 1000);
+      return sleep(1000)
+        .then(() => {
+          touchFile(watchedFile);
+          return sleep(1000);
+        })
+        .then(() => {
+          mocha.kill('SIGINT');
+          return outputPromise;
+        })
+        .then(results => {
+          expect(results, 'to have length', 2);
+        });
+    });
+
+    it('ignores files in "node_modules" and ".git"', function() {
+      const nodeModulesFile = path.join(
+        this.tempDir,
+        'node_modules',
+        'file.xyz'
+      );
+      const gitFile = path.join(this.tempDir, '.git', 'file.xyz');
+
+      touchFile(gitFile);
+      touchFile(nodeModulesFile);
+
+      const [mocha, outputPromise] = runMochaJSONWatchAsync(
+        this.testFile,
+        ['--extension', 'xyz,js'],
+        {
+          cwd: this.tempDir
+        }
+      );
+
+      return sleep(1000)
+        .then(() => {
+          touchFile(gitFile);
+          touchFile(nodeModulesFile);
+        })
+        .then(() => sleep(1000))
+        .then(() => {
+          mocha.kill('SIGINT');
+          return outputPromise;
+        })
+        .then(results => {
+          expect(results, 'to have length', 1);
+        });
     });
   });
 });
+
+/**
+ * Invokes the mocha binary with the `--watch` argument for the given fixture.
+ *
+ * Returns child process and a promise for the test results. The test results
+ * are an array of JSON objects generated by the JSON reporter.
+ *
+ * Checks that the exit code of the mocha command is 130, i.e. mocha was killed
+ * by SIGINT.
+ */
+function runMochaJSONWatchAsync(fixture, args, spawnOpts) {
+  args = ['--watch'].concat(args);
+  const [mocha, mochaDone] = runMochaJSONRawAsync(fixture, args, spawnOpts);
+  const testResults = mochaDone.then(data => {
+    expect(data.code, 'to be', sigintExitCode);
+
+    const testResults = data.output
+      // eslint-disable-next-line no-control-regex
+      .replace(/\u001b\[\?25./g, '')
+      .split('\u001b[2K')
+      .map(x => JSON.parse(x));
+    return testResults;
+  });
+  return [mocha, testResults];
+}
+
+/**
+ * Touch a file by appending a space to the end. Returns a promise that resolves
+ * when the file has been touched.
+ */
+function touchFile(file) {
```

Comment:
this function doesn't return anything.

Reply:
Updated the documentation in https://github.com/mochajs/mocha/pull/3915/commits/3948056608857728f63732b7f6a11286a5cf0447